File: flutter/lib/data/repository/ai/ai_prompt_builder_repository_impl.dart
import 'package:dart_openai/dart_openai.dart';

import '../../../domain/repository/ai/ai_prompt_builder_repository.dart';

class AIPromptBuilderRepositoryImpl implements AIPromptBuilderRepository
<OpenAIChatCompletionChoiceMessageModel, List<OpenAIChatCompletionChoiceMessageModel>> {
  @override
  late OpenAIChatCompletionChoiceMessageModel systemPrompt;

  @override
  late OpenAIChatCompletionChoiceMessageModel userPrompt;

  @override
  void setSystemPrompt(String prompt) {
    systemPrompt = OpenAIChatCompletionChoiceMessageModel(
      role: OpenAIChatMessageRole.system,
      content: [
        OpenAIChatCompletionChoiceMessageContentItemModel.text(
            prompt
        ),
      ],
    );
  }

  @override
  void setUserPrompt(String prompt) {
    userPrompt = OpenAIChatCompletionChoiceMessageModel(
      role: OpenAIChatMessageRole.user,
      content: [
        OpenAIChatCompletionChoiceMessageContentItemModel.text(
            prompt
        ),
      ],
    );
  }

  @override
  buildPrompt() {
    return [systemPrompt, userPrompt];
  }

}

File: android/app/src/main/java/com/synergyboat/flashcardAi/data/repository/ai/AiPromptBuilderRepositoryImpl.kt
package com.synergyboat.flashcardAi.data.repository.ai

import com.aallam.openai.api.chat.ChatCompletionRequest
import com.aallam.openai.api.chat.ChatMessage
import com.aallam.openai.api.chat.ChatRole
import com.aallam.openai.api.model.ModelId
import com.synergyboat.flashcardAi.core.Constants
import com.synergyboat.flashcardAi.domain.repository.ai.AiPromptBuilderRepository
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class AiPromptBuilderRepositoryImpl @Inject constructor(): AiPromptBuilderRepository<ChatCompletionRequest> {
    override fun buildPrompt(topic: String, count: Int): ChatCompletionRequest {
        return ChatCompletionRequest(
            model = ModelId("gpt-3.5-turbo"),
            messages = listOf(
                ChatMessage(
                    role = ChatRole.System,
                    content = Constants.PROMPT
                ),
                ChatMessage(
                    role = ChatRole.User,
                    content = "Generate flashcards for the topic: '$topic'.\n Number of cards to generate: '$count'."
                )
            )
        )
    }
}